### ISSUE TITLE
Regex breaks whole word search

### ISSUE THREAD

**Issue by guusw** [Microsoft/vscode#23785]
- VSCode Version: Code 1.9.1 (f9d0c687ff2ea7aabd85fb9a43129117c0ecf519, 2017-02-08T23:44:55.542Z)
- OS Version: Linux x64 4.9.8-1-ARCH
- Extensions:

|Extension|Author|Version|
|---|---|---|
|cpptools|ms-vscode|0.10.2|
|csharp|ms-vscode|1.7.0|
|cmake|twxs|0.0.15|
|cmake-tools|vector-of-bool|0.9.4|

---

Using negative lookahead on file search seems to break whole word search

For example:
 String(?!\.hpp)
seems to pick up words like StringLocation (although whole word search is on)

**Comment by Spown** [comment]
Well, you see, you have a rare case of a conflict of interests here. "whole word search" isn't a part of RegEx mechanic - it is another layer search refinement of the VSCode itself, that is applied __after__ the regEx search. So basically when you use ``String(?!.hpp)`` you basically say "find me every occurrence of "String" that doesn't have ".hpp" immediately after; then filter only whole words" - you doesn't say "find me whole words that start with "String" that doesn't have ".hpp" immediately after". So RegEx deos exactly what you told it to do: it finds every "String" that doesn't have ".hpp" and then removes every non-whole word match... but wait, at this point every RegEx fund is a whole word on its own even if in the whole text it is a part of a word.

Basically you have to forget whole word search even exists when using RegEx and try to encapsulate all your needs in one pattern, like this ``String(?!\.hpp)(\w*)?`` (if I get what you need):
![](https://drive.google.com/uc?id=0B-OJBofmehO3OWtQS01KSURnNmc)

Also VSCode 1.10.2 is the current ver. Update maybe?

**Comment by roblourens** [comment]
Right, this is a weakness with how we implement whole word search, and I'm not sure how to fix it.

It is actually implemented partly with the regex engine itself, using the `\b` "word boundary" class. But, we only add `\b` if the search query doesn't end in a word boundary already: https://github.com/Microsoft/vscode/blob/master/src/vs/base/common/strings.ts#L211. This just makes it work a little more like you'd expect, like recently came up in this issue: https://github.com/Microsoft/vscode/issues/23534#issuecomment-290245659

The problem is in a case like this, your query ends in `)`, but that's part of the regex. So maybe we shouldn't have made it so clever, because I don't think it's possible to fix it in general. You can have a regex that sometimes will match a `\B` at the end, and sometimes won't, so we can't just always add `\b` or not for a particular query.

**Comment by roblourens** [comment]
I think we could fix it by changing our current logic to use a pattern that either matches a lookbehind for `\b`, or matches `\b`. Something like `((?<=\b)|\b)`. That doesn't work, `(?<=\b)` doesn't match when I think it should, but maybe something like that could work.

But another problem is that I'd want the same logic for text search, which now doesn't support lookaround as of 1.11 after we switched to ripgrep.

**Comment by Spown** [comment]
I would suggest disabling whole word search gui switcher when RegExp is enabled to not give people false hopes.

**Comment by roblourens** [comment]
`\b` matches between a word and non-word character, but we want something like that which also matches between any two non-word characters. Appending `(\b|\W)` would do that, but then the `\W` character becomes part of the match. And again, no lookaround. We could post process all the matches to determine whether we need to slice a character off. Or, run it with the basic pattern and post process to determine whether a match is for a 'whole word'. I'm not enthusiastic about either of those options.
